pkg/handler/handler.go

```diff
@@ -83,50 +114,40 @@ func NewOpenAPIService(spec *spec.Swagger) (*OpenAPIService, error) {
 	return o, nil
 }
 
-func (o *OpenAPIService) getSwaggerBytes() ([]byte, string, time.Time) {
+func (o *OpenAPIService) getSwaggerBytes() ([]byte, string, time.Time, error) {
 	o.rwMutex.RLock()
 	defer o.rwMutex.RUnlock()
-	return o.specBytes, o.specBytesETag, o.lastModified
-}
-
-func (o *OpenAPIService) getSwaggerPbBytes() ([]byte, string, time.Time) {
-	o.rwMutex.RLock()
-	defer o.rwMutex.RUnlock()
-	return o.specPb, o.specPbETag, o.lastModified
+	specBytes, etag, err := o.jsonCache.Get()
+	if err != nil {
+		return nil, "", time.Time{}, err
+	}
+	return specBytes, etag, o.lastModified, nil
 }
 
-func (o *OpenAPIService) getSwaggerPbGzBytes() ([]byte, string, time.Time) {
+func (o *OpenAPIService) getSwaggerPbBytes() ([]byte, string, time.Time, error) {
 	o.rwMutex.RLock()
 	defer o.rwMutex.RUnlock()
-	return o.specPbGz, o.specPbGzETag, o.lastModified
-}
-
-func (o *OpenAPIService) UpdateSpec(openapiSpec *spec.Swagger) (err error) {
-	specBytes, err := json.Marshal(openapiSpec)
+	specPb, etag, err := o.protoCache.Get()
 	if err != nil {
-		return err
+		return nil, "", time.Time{}, err
 	}
-	specPb, err := ToProtoBinary(specBytes)
-	if err != nil {
-		return err
-	}
-	specPbGz := toGzip(specPb)
-
-	specBytesETag := computeETag(specBytes)
-	specPbETag := computeETag(specPb)
-	specPbGzETag := computeETag(specPbGz)
-
-	lastModified := time.Now()
+	return specPb, etag, o.lastModified, nil
+}
 
+func (o *OpenAPIService) UpdateSpec(openapiSpec *spec.Swagger) (err error) {
 	o.rwMutex.Lock()
 	defer o.rwMutex.Unlock()
-
-	o.specBytes = specBytes
-	o.specPb = specPb
-	o.specPbGz = specPbGz
-	o.specBytesETag = specBytesETag
-	o.specPbETag = specPbETag
-	o.specPbGzETag = specPbGzETag
+	o.jsonCache = o.jsonCache.New(func() ([]byte, error) {
+		return json.Marshal(openapiSpec)
+	})
+	o.protoCache = o.protoCache.New(func() ([]byte, error) {
+		json, _, err := o.jsonCache.Get()
+		if err != nil {
+			return nil, err
+		}
+		return ToProtoBinary(json)
+	})
+	lastModified := time.Now()
 	o.lastModified = lastModified
```

Comment:
why do we create `lastModified` here?
```suggestion
	o.lastModified = time.Now()
```

Reply:
Hi there! Sorry for jumping in. Was wondering won't this `lastModified` be evaluated independently on each api-server replica and result in different results? If so - can this break `Not Modified` responses from `http.ServeContent`?

Reply:
@apelisse Good question - I think it was left over from an earlier refactoring, and I missed cleaning it up, thanks for spotting that!

@nekto0n This is also a good question; I'm not sure what the answer is, so I hope someone else can jump in. As far as I know, the last-modified time was calculated like this before this PR, so it's consistent with previous behavior at least. It's another question entirely if that behavior was correct in the first place though, haha.

Reply:
I think that's fine. `lastModified` is used to evaluate the freshness of the cache, but never its correctness (we use the etag for that). We don't really use the freshness anywhere, but even if we did I don't think this would necessarily be a problem since it doesn't have to be exact like the etag does.